src/compiler/checker.ts

```diff
@@ -6988,8 +6983,9 @@ namespace ts {
                 getPropertiesOfObjectType(type);
         }
 
-        function isTypeInvalidDueToUnionDiscriminant(contextualType: Type, obj: ObjectLiteralExpression): boolean {
-            return obj.properties.some(property => {
+        function isTypeInvalidDueToUnionDiscriminant(contextualType: Type, obj: ObjectLiteralExpression | JsxAttributes): boolean {
+            const list = obj.properties as NodeArray<ObjectLiteralElementLike | JsxAttributeLike>;

```

Comment:
can ObjectLiteralElementLike be expanded to include JsxAttributeLike, or could we introduce yet another union to unify the two?

Reply:
I only need to perform this cast because we can't resolve signatures of unions, so operating on a union of two arrays is impossible.